### PR TITLE
Fix test failures on master

### PR DESCRIPTION
**Pull request by d-chambers**
### What does this PR do?

Fixes the test failures on master due to flake8, escape characters in python 3.7, and a few other misc. issues.

### Why was it initiated?  Any relevant Issues?

The test suite on master was failing due to several small issues. 

### PR Checklist
- [x] Correct base branch selected? `master` for new features, `maintenance_...` for bug fixes
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [ ] If the PR is making changes to documentation, docs pages can be built automatically.
      Just remove the space in the following string after the + sign: "+ DOCS"
- [x] If any network modules should be tested for the PR, add them as a comma separated list
      (e.g. `clients.fdsn,clients.arclink`) after the colon in the following magic string: "+TESTS:"
      (you can also add "ALL" to just simply run all tests across all modules)
- [X] All tests still pass.
- [ ] Any new features or fixed regressions are be covered via new tests.
- [ ] Any new or changed features have are fully documented.
- [ ] Significant changes have been added to `CHANGELOG.txt` .
- [ ] First time contributors have added your name to `CONTRIBUTORS.txt` .

### CODE REVIEW

**Comment by d-chambers** [comment]
Well it works on my machine 😉

**Comment by calum-chamberlain** [comment]
+TESTS:ALL

**Comment by calum-chamberlain** [comment]
I have done a few things in here to make some failing network tests pass - these all seemed to be due to changes in sensor availability, or changes in how the service was delivered. 

I haven't fixed a failing fdsn test `obspy.clients.fdsn.tests.ClientTestCase.test_dataselect_bulk` which checks for a few things, including the use of the `minimumlength` argument.  To me it looks like how IRIS handle that might have changed.  The test specifies a start and end time 4s apart and a `minimumlength` of 5. We are now getting traces less than minimum length. In the FDSN docs it isn't clear whether `starttime` and `endtime` or `minimumlength` should take priority.

For the molepeople plot issues, I did some experimentation and it looks like `proj4=5.2.0` is causing the issue - I tested 4.9.3 and 5.1.0 and found both to work.  I have added a check of the version when importing `obspy.imaging.maps` which follows the other warnings.  I also pinned the version for ci to 4.9.3, this probably isn't ideal.

**Comment by calum-chamberlain** [comment]
I'm not sure what is going on with the appveyor fail relating to the use of `TemporaryWorkingDirectory` - previously I added a check to make sure that the files were closed (they are) but it still complains that something else is accessing the files.  Not sure if we should just ignore it?

**Comment by d-chambers** [comment]
> I also pinned the version for ci to 4.9.3, this probably isn't ideal.

Yes, looks like python 3.7 cant use proj4 4.9.3:
```
UnsatisfiableError: The following specifications were found to be in conflict:
  - proj4=4.9.3
  - pyproj
  - python=3.7
Use "conda info <package>" to see the dependencies for each package.
```

Maybe we should just skip the test if the version is greater than 5.1 for most builds and only pin it on one/a few of  them?

**Comment by calum-chamberlain** [comment]
I have marked one of the network tests as an expected fail just to see what other CI fails we have. It fails at the moment because `minimumlength` looks like it is being ignored.  I'm not sure if IRIS changed how they respond to requests or if it is on our end? The following example shows the issue:
```python
from obspy import UTCDateTime
from obspy.clients.fdsn import Client

client = Client("IRIS")
st = client.get_waveforms(
    station="A25A", network="TA", channel="BHZ", location="*",
    starttime=UTCDateTime("2011-07-22T14:50:23"),
    endtime=UTCDateTime("2011-07-22T14:50:29"), minimumlength=5)
print(st)
"""
Returns:
1 Trace(s) in Stream:
TA.A25A..BHZ | 2011-07-22T14:50:23.000000Z - 2011-07-22T14:50:25.500000Z | 40.0 Hz, 101 samples
which is less than 5 seconds long.
"""
```

**Comment by calum-chamberlain** [comment]
Current failures: 
1. There are still a few plotting issues which seem to be related to matplotlib version 3. The [PPSD test fail](https://travis-ci.org/obspy/obspy/jobs/463035988#L2843) is just a change in where xticks are plotted for version 3 (this also happens on the appveyor run). 
2. I'm not sure what the [basemap related fails](https://travis-ci.org/obspy/obspy/jobs/463035988#L2707) are on 3.7
3. Some [taup fails](https://travis-ci.org/obspy/obspy/jobs/463035991#L6160) on MacOS Python 3.7 that I haven't debugged.
4. A [simulate fail on appveyor](https://ci.appveyor.com/project/obspy/obspy/builds/20742130/job/alxfomlq6cl4ierq#L963), that I haven't looked into yet.

I'm going to focus on other things for a bit - @d-chambers do you want to do anything more on this?

File: appveyor.yml [hunk 1]
```diff
@@ -113,7 +113,7 @@ install:
   - "activate test"
   # Install default dependencies
   # pip 10.0.1 seems buggy and breaks all our appveyor builds
-  - "conda install -q --yes \"pip<10\" numpy scipy matplotlib lxml sqlalchemy mock nose gdal decorator requests basemap jsonschema pyshp cryptography"
+  - "conda install -q --yes \"pip<10\" numpy scipy matplotlib lxml sqlalchemy mock nose gdal decorator requests basemap jsonschema pyshp cryptography \"proj4=4.9.3\""
```
Comment (calum-chamberlain) [inline appveyor.yml]: Could unpin proj4 here, but that would likely mean that plotting of maps wouldn't get testes on windows.

File: obspy/clients/fdsn/tests/test_client.py [hunk 2]
```diff
@@ -668,6 +669,7 @@ def test_str_method(self):
         expected = normalize_version_number(expected)
         self.assertEqual(got, expected, failmsg(got, expected))
 
+    @unittest.expectedFailure
```
Comment (calum-chamberlain) [inline obspy/clients/fdsn/tests/test_client.py]: Does anyone know if IRIS has changed how they respond to `minimumlength`?

**Comment by d-chambers** [comment]
I should have time to pick this up again after AGU, I will take a look at the remaining issues at that point.

File: obspy/clients/fdsn/tests/test_client.py (hunk 2 above)
Comment (d-chambers) [reply obspy/clients/fdsn/tests/test_client.py]: Maybe @chad-iris or @nick-iris know or could find out for us?

**Comment by megies** [comment]
> It fails at the moment because `minimumlength` looks like it is being ignored. I'm not sure if IRIS changed how they respond to requests or if it is on our end? The following example shows the issue:

Looks like a server side issue at IRIS. I'll drop @chad-iris a short mail, he can forward it to the person in charge.

> Could unpin proj4 here, but that would likely mean that plotting of maps wouldn't get tested on windows.

Actually these package version specs in the CI yamls have been annoying me for a long while. I think at some point it would be best to move the packages specs from the yamls to conda pinning files that we copy in the conda test env during CI. That would also make it much easier to avoid getting different package versions when doing more installation commands subsequently. But this should be discussed in a separate issue first, I think.

>     2\. I'm not sure what the [basemap related fails](https://travis-ci.org/obspy/obspy/jobs/463035988#L2707) are on 3.7

That the `..hold` thingy? This is a known issue of some matplotlib versions that broke basemap, I can't find the corresponding issue right now unfortunately..

**Comment by megies** [comment]
@d-chambers @calum-chamberlain if you don't have time to work on the rest of the failures right now, I would still suggest we merge this as soon as possible, whenever appropriate. Best not keep this hanging and get it merged soonish, even if some fails remain.

Feel free to merge when you feel like it!

**Comment by calum-chamberlain** [comment]
The `axis._hold` type issue is [related to this](https://github.com/matplotlib/basemap/issues/435) I think. Added a `skipIf` for matplotlib version 3.0.1 and basemap >= 1.1.0.
I can't reproduce the appveyor `test_simulate` fail, but it looks like it might be similar to #2188. I couldn't reproduce on my 64Bit windows with a clean conda env using the same packages as appveyor.

**Comment by d-chambers** [comment]
> I would still suggest we merge this as soon as possible

Good idea. I wont have time to do anything on this for a week or two so go ahead and merge whenever you are happy @calum-chamberlain. It is nice to see Travis green again.

**Comment by megies** [comment]
> We could just check `allclose` on data on windows as was done for #2188?

[This the only info we have on the fail?](https://ci.appveyor.com/project/obspy/obspy/builds/20776339/job/1589igujyjgducwi#L952)

```
Traceback (most recent call last):
  File "C:\projects\obspy\obspy\core\tests\test_stream.py", line 2207, in test_simulate_seedresp_parser
    self.assertEqual(tr1, tr2)
AssertionError: <obspy.core.trace.Trace object at 0x00000002A2D67080> != <obspy.core.trace.Trace object at 0x00000002A66E6438>
```

If that's the case, we should add an `assertEqual` on the `tr.data` to see how big the difference is, first of all?



> skipping tests for specific versions that either do not work (such as basemap with mpl 3.0.1,

this is the right thing to do, IMHO. we won't see a lot of mpl 3.0.1 anyway once the next minor release is out (that'll fix this hold issue)


> skipping the PPSD plot test for mpl>=3 which moves the xtick locations - I think the new locations are nicer and result in a better use of the figure space

At some point we want to update our baseline images to mpl 3. Maybe we want to wait for mpl 3.0.2, though. We can then have higher tolerances for failing tests with mpl 2.x
I'm okay with merging this right now, just open an issue please as a reminder to switch baseline images to mpl 3 at some point, ideally against the 1.2.0 milestone?

> And ignoring the network test change for `minimumlength`.

Hmm. not sure about this one.. maybe better leave it in as a failing test? I strongly believe it's a server side bug that will be fixed pretty fast by IRIS once it gets the attention of the person in charge (I dropped Chad a mail to forward the issue but he didn't respond yet, probably busy with AGU preps).



> I'm also not super keen on the change I made to [`taup`](https://github.com/obspy/obspy/pull/2269/commits/7eabcca746e89335984e548745a2a5b61b28d3f6#diff-59955951fd157f214a342cd6e6160b84L156) - changing from resizing inplace to generating a new array with `np.resize` as a function might have implications for memory and speed, but I don't know the taup stuff well enough to know if the size of these arrays is going to be an issue

This looks like we need to be careful not to break things.. dunno. I also don't know the taup stuff much, but this looks like it can break references to the same data array?

**Comment by calum-chamberlain** [comment]
Thanks for those comments @megies. 
- I have removed the changes to `obspy.taup.tau_branch`, so I expect those doctests to fail. I think the fails could be related to how the doctests are run - maybe each doctest is trying to work on the same array at the same time? 
- Removed the expected fail flag on the test with the `minimumlength` issue.
- Added some (hopefully useful) extra comparisons for the two `simulate` tests that seemed flakey, hopefully we can see whether there really is an issue or not with that.

**Comment by chad-iris** [comment]
>  It fails at the moment because minimumlength looks like it is being ignored. I'm not sure if IRIS changed how they respond to requests or if it is on our end?

Indeed the option was effectively being ignored.  This is now fixed at IRIS.

**Note**: We are deprecating this option as we have a much better way to do this with our Research Ready Data Sets service, which allows selection of data based on quality metrics including time coverage.  It's supported for now, but not promised to be working in the future.

It's been broken for 6 months and this is the first we've heard about it, which aligns with my suspicion that it's not often used.

**Comment by calum-chamberlain** [comment]
@chad-iris thanks! Personally I haven't used the `minimumlength` constraint, and wouldn't have known it wasn't working if not for the tests.  Nice to hear about the RRDS too, I wasn't aware of that.
Sounds like it might be a good idea for us to not base a test on this then if it won't be supported in the future.

**Comment by calum-chamberlain** [comment]
The simulate fail that remains on windows passes the tests comparing stats and using `allclose` (with default settings, so `rtol=1e-7`) but still fails the [stream and trace comparison tests](https://ci.appveyor.com/project/obspy/obspy/builds/20828146/job/fkbjsw88toc97w7p#L941)

**Comment by megies** [comment]
> Sounds like it might be a good idea for us to not base a test on this then if it won't be supported in the future.

Without looking at the (test) code, we could replace the test by checking that the appropriate URL is built inside the client. But then again, I am confident that the URL building inside the client is tested thoroughly enough, so probably we can just get rid of that parameter when downloading data in tests. On the other hand, since it's fixed on server side, we can also just leave it as is for now.

**Comment by d-chambers** [comment]
I prefer the second option. It should also have an option to ignore `processing` attribute of the `stats` object, as that has been a sticking point in the past as well. Although this would be a nice feature, it probably belongs in a separate PR no?

**Comment by megies** [comment]
OK sounds good, so probably add a utility Stream/Trace assertion function in `core.util.testing` but postpone for a future PR.

**Comment by megies** [comment]
In any case, I'm gonna merge this now to not have it in limbo for too long. Thanks for the initiative on this tedious matter @d-chambers and @calum-chamberlain.

**Comment by megies** [comment]
We might decide to backport this for 1.1.1 to get clean CI builds, probably, see #1983

**Pull request merged**